sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BeamFnDataReadRunnerTest.java

```diff
@@ -122,334 +129,579 @@
-            valueInGlobalWindow("C"),
-            valueInGlobalWindow("D"),
-            valueInGlobalWindow("E")));
+    @Parameterized.Parameter(0)
+    public ProcessBundleSplitResponse expectedResponse;
+
+    @Parameterized.Parameter(1)
+    public long inputElements;
+
+    @Parameterized.Parameter(2)
+    public List<String> processedElements;
+
+    @Parameterized.Parameter(3)
+    public double fractionOfRemainder;
+
+    @Parameterized.Parameter(4)
+    public List<Long> allowedSplitPoints;
+
+    @Test
+    public void testChannelSplittingWithAllowedSplitPoints() throws Exception {
+      List<WindowedValue<String>> outputValues = new ArrayList<>();
+      BeamFnDataClient mockBeamFnDataClient = mock(BeamFnDataClient.class);
+      BeamFnDataReadRunner<String> readRunner =
+          createReadRunner(outputValues::add, PTRANSFORM_ID, mockBeamFnDataClient);
+      assertEquals(
+          expectedResponse,
+          executeSplit(
+              readRunner,
+              PTRANSFORM_ID,
+              inputElements,
+              processedElements,
+              fractionOfRemainder,
+              allowedSplitPoints));
+    }
   }
 
-  @Test
-  public void testSplittingWhenSomeElementsProcessed() throws Exception {
-    List<WindowedValue<String>> outputValues = new ArrayList<>();
-    BeamFnDataReadRunner<String> readRunner = createReadRunner(outputValues::add);
-
-    ProcessBundleSplitRequest request =
-        ProcessBundleSplitRequest.newBuilder()
-            .putDesiredSplits(
-                "pTransformId",
-                DesiredSplit.newBuilder()
-                    .setEstimatedInputElements(10)
-                    .setFractionOfRemainder(0.5)
-                    .build())
-            .build();
-    ProcessBundleSplitResponse.Builder responseBuilder = ProcessBundleSplitResponse.newBuilder();
-
-    // Process 2 elements then split
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("A"));
-    readRunner.forwardElementToConsumer(valueInGlobalWindow("B"));
-    readRunner.trySplit(request, responseBuilder);
+  // Test different cases of element split with empty allowed split points.
+  @RunWith(Parameterized.class)
+  public static class ElementSplitTest {
+    @Parameterized.Parameters
+    public static Iterable<Object[]> data() {
+      return ImmutableList.<Object[]>builder()
+          // Split between future elements at element boundaries.
+          .add(new Object[] {channelSplitResult(2L), 4L, ImmutableList.of("A"), 0.51, 0})
+          .add(new Object[] {channelSplitResult(2L), 4L, ImmutableList.of("A"), 0.49, 0})
+          .add(new Object[] {channelSplitResult(1L), 4L, ImmutableList.of("A"), 0.26, 0})
+          .add(new Object[] {channelSplitResult(1L), 4L, ImmutableList.of("A"), 0.25, 0})
+          // If the split falls inside the first, splittable element, split there.
+          .add(new Object[] {elementSplitResult(0L), 4L, ImmutableList.of("A"), 0.2, 0})
```

Comment:
It would be nice if the fraction did make it through into the split result allowing us to validate correctly computing the fraction.